File: ios/LivenessCameraPlugin.swift
import VisionCamera
import MLKitFaceDetection
import MLKitVision
import UIKit
import AVFoundation

@objc(LivenessCameraPlugin)
public class LivenessCameraPlugin: FrameProcessorPlugin {

  private var faceDetector: FaceDetector

  // VC v4.5+ removed VisionCameraProxyHolder — use the no-arg init instead.
  public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]? = [:]) {
    let opts = FaceDetectorOptions()
    opts.performanceMode = .fast
    opts.classificationMode = .all
    opts.landmarkMode = .none
    opts.contourMode = .none
    opts.isTrackingEnabled = false
    self.faceDetector = FaceDetector.faceDetector(options: opts)
    super.init(proxy: proxy, options: options)
  }

  public override func callback(
    _ frame: Frame,
    withArguments arguments: [AnyHashable: Any]?
  ) -> Any {
    let image = VisionImage(buffer: frame.buffer)
    image.orientation = uiOrientation(from: frame)

    do {
      let faces = try faceDetector.results(in: image)
      guard let face = faces.first else {
        return ["detected": false]
      }

      // face.frame is in the coordinate space of the INPUT buffer (before rotation).
      // Report the buffer dimensions so JS can pick the right reference axis.
      let bufW = Double(CVPixelBufferGetWidth(frame.buffer))
      let bufH = Double(CVPixelBufferGetHeight(frame.buffer))

      return [
        "detected": true,
        "bounds": [
          "x": Double(face.frame.origin.x),
          "y": Double(face.frame.origin.y),
          "width": Double(face.frame.size.width),
          "height": Double(face.frame.size.height),
        ],
        "frameWidth": bufW,
        "frameHeight": bufH,
        "yawAngle": Double(face.headEulerAngleY),
        "pitchAngle": Double(face.headEulerAngleX),
        "rollAngle": Double(face.headEulerAngleZ),
        "leftEyeOpenProbability": Double(face.leftEyeOpenProbability),
        "rightEyeOpenProbability": Double(face.rightEyeOpenProbability),
        "smilingProbability": Double(face.smilingProbability),
      ]
    } catch {
      return ["detected": false]
    }
  }

  // MARK: - Orientation helper

  /// Map Vision Camera's frame orientation to the UIImage.Orientation that
  /// ML Kit needs so it sees an upright face regardless of device rotation.
  private func uiOrientation(from frame: Frame) -> UIImage.Orientation {
    switch frame.orientation {
    case .portrait:           return .right
    case .portraitUpsideDown: return .left
    case .landscapeLeft:      return .up
    case .landscapeRight:     return .down
    @unknown default:         return .right
    }
  }
}

File: android/src/main/java/com/livenesscamera/LivenessCameraPlugin.kt
package com.livenesscamera

import android.media.Image
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin

class LivenessCameraPlugin : FrameProcessorPlugin() {

  private val faceDetector = FaceDetection.getClient(
    FaceDetectorOptions.Builder()
      .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
      .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
      .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
      .setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
      .build()
  )

  /**
   * Convert Vision Camera's Orientation to ML Kit rotation degrees.
   * Uses toString() to stay resilient across VC patch-version API churn.
   */
  private fun orientationDegrees(frame: Frame): Int {
    val name = frame.orientation.toString().uppercase()
    return when {
      name.contains("LANDSCAPE_LEFT") -> 90
      name.contains("PORTRAIT_UPSIDE_DOWN") -> 180
      name.contains("LANDSCAPE_RIGHT") -> 270
      else -> 0 // PORTRAIT
    }
  }

  override fun callback(frame: Frame, arguments: Map<String, Any>?): Any {
    val mediaImage: Image = frame.image
    val inputImage = InputImage.fromMediaImage(mediaImage, orientationDegrees(frame))

    return try {
      // Tasks.await blocks the frame-processor thread synchronously.
      // ML Kit face detection is fast (~5–10 ms) so this is acceptable.
      val faces = Tasks.await(faceDetector.process(inputImage))

      if (faces.isEmpty()) {
        return mapOf("detected" to false)
      }

      val face = faces.first()
      val box = face.boundingBox

      // IMPORTANT: All numeric values must be Double, not Float.
      // JSI (Vision Camera's JS bridge) cannot convert Java Float → jsi::Value
      // and throws "Cannot convert Java type class java.lang.Float" at runtime.
      mapOf(
        "detected" to true,
        "bounds" to mapOf(
          "x" to box.left.toDouble(),
          "y" to box.top.toDouble(),
          "width" to box.width().toDouble(),
          "height" to box.height().toDouble()
        ),
        // Pass both buffer dimensions so JS can use the shorter side as the
        // face-size reference (avoids landscape-vs-portrait ratio errors).
        "frameWidth" to mediaImage.width.toDouble(),
        "frameHeight" to mediaImage.height.toDouble(),
        "yawAngle" to face.headEulerAngleY.toDouble(),
        "pitchAngle" to face.headEulerAngleX.toDouble(),
        "rollAngle" to face.headEulerAngleZ.toDouble(),
        "leftEyeOpenProbability" to (face.leftEyeOpenProbability?.toDouble() ?: -1.0),
        "rightEyeOpenProbability" to (face.rightEyeOpenProbability?.toDouble() ?: -1.0),
        "smilingProbability" to (face.smilingProbability?.toDouble() ?: -1.0)
      )
    } catch (e: Exception) {
      mapOf("detected" to false)
    }
  }
}
